beam-sync.md

```diff
@@ -0,0 +1,187 @@
+only serve state for recent blocks, configured differently for different clients. A
+common number is ~100 blocks of state retained. If you fall behind this, then missing state
+may no longer be available to you. This can leave you indefinitely stuck, so when we
+lag far enough behind, we pivot.
+
+### Pivoting
+
+Pivoting skips executing the current blocks, and chooses the tip block again. It will
+typically increase the missing state that must be backfilled. It is quite likely that Beam Sync
+v0 will require a few pivots before it exits. As it progresses, and stores more state, pivoting gets less likely.
+v1 and v2 should be fast enough that pivoting is just a failsafe against "attack blocks" with
+intentionally gigantic witnesses.
+
+### Optimizations
+
+A naive implementation of Beam Sync, given that there are on the order of 4k trie nodes read per mainnet block, and a good round-trip time is ~100ms, it would take 400s to collect all the required data to execute a single block. With ~15-second block times, this is clearly a problem.
+
+So while optimizations are not *technically* required, they are necessary in practice to avoid getting stuck in a loop of pivots. A pivot loop would mean never catching up to complete Full Sync.
+
+Below are some optimizations that Trinity has found effective. You may notice the theme that they all provide ways to look up more data in parallel. We would much rather be limited by total peer bandwidth across all connected peers, rather than limited by the round-trip latency of a single peer.
+
+#### Optimization 1: Address Batching
+
+Before starting EVM execution, we can identify several key addresses that will be needed during the block: transaction recipients, transaction senders, and the header's coinbase.
+
+Critically, the trie nodes for these addresses can be collected in parallel. Rather than requesting a single trie node at a time, you can walk down one layer of the trie for all the addresses at once. Say you have ~200 addresses to collect, your `GetNodeData` might include up to 200 trie nodes at once on the final layer. This is a dramatic improvement, lessening the sensitivity to peer latency.
+
+#### Optimization 2: Parallel Execution
+
+The flow diagram above shows block imports as serial, so that child blocks aren't executed until the parent is complete. We want to improve on that with Parallel Execution. We start executing child blocks before the parent has completed executing. Some implications:
+
+- The JIT-data limit of only requesting one trie node at a time applies per-block, so they can be requested in parallel across all pending blocks.
+- The block count lag determines how many pending blocks are available. If you are right at the tip, parallelism drops to 1. The benefit of previewing grows as you lag and shrinks as you catch up. It is great to catch up faster, the further you are behind, but it can make Beam Sync feel a bit "stutter-y".
```

Comment:
maybe replace "stuttery" with an explicit explanation of "falling behind and then quickly catching up only to fall behind again..."